File: app/src/main/java/com/tomclaw/appsend/util/BubbleColorDrawable.java
package com.tomclaw.appsend.util;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PixelFormat;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.drawable.Drawable;

import androidx.annotation.NonNull;

/**
 * Created by solkin on 13.07.15.
 */
public class BubbleColorDrawable extends Drawable {

    private static int OFFSET;
    private static int BUBBLE_RADIUS;
    private final int color;
    private final Corner corner;
    private Paint whitePaint;

    public BubbleColorDrawable(Context context, int color, Corner corner) {
        this.color = color;
        this.corner = corner;
        OFFSET = dp(8, context);
        BUBBLE_RADIUS = dp(6, context);
        whitePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    }

    @Override
    public void setColorFilter(ColorFilter cf) {

    }

    @Override
    public void setAlpha(int alpha) {

    }

    @Override
    public int getOpacity() {
        return PixelFormat.OPAQUE;
    }

    @Override
    public void draw(@NonNull Canvas canvas) {
        Rect r = getBounds();
        RectF rect = new RectF(r);
        rect.inset(OFFSET, OFFSET);
        Path path = new Path();
        if (corner == Corner.LEFT) {
            path.moveTo(0, 0);
            path.lineTo(r.width() - OFFSET, 0);
            path.arcTo(new RectF(r.right - BUBBLE_RADIUS * 2, 0, r.right, BUBBLE_RADIUS * 2), 270, 90);
            path.lineTo(r.width(), r.height() - OFFSET);
            path.arcTo(new RectF(r.right - BUBBLE_RADIUS * 2, r.bottom - BUBBLE_RADIUS * 2, r.right, r.bottom), 0, 90);
            path.lineTo(BUBBLE_RADIUS, r.height());
            path.arcTo(new RectF(OFFSET, r.bottom - BUBBLE_RADIUS * 2, BUBBLE_RADIUS * 2 + OFFSET, r.bottom), 90, 90);
            path.lineTo(OFFSET, OFFSET);
        } else if (corner == Corner.RIGHT) {
            path.moveTo(BUBBLE_RADIUS, 0);
            path.lineTo(r.width(), 0);
            path.lineTo(r.width() - OFFSET, OFFSET);
            path.lineTo(r.width() - OFFSET, r.height() - BUBBLE_RADIUS);
            path.arcTo(new RectF(r.right - BUBBLE_RADIUS * 2 - OFFSET, r.bottom - BUBBLE_RADIUS * 2, r.right - OFFSET, r.bottom), 0, 90);
            path.lineTo(BUBBLE_RADIUS, r.height());
            path.arcTo(new RectF(0, r.bottom - BUBBLE_RADIUS * 2, BUBBLE_RADIUS * 2, r.bottom), 90, 90);
            path.lineTo(0, BUBBLE_RADIUS);
            path.arcTo(new RectF(0, 0, BUBBLE_RADIUS * 2, BUBBLE_RADIUS * 2), 180, 90);
        } else if (corner == Corner.NONE) {
            rect.inset(-OFFSET, -OFFSET);
            path.addRoundRect(rect, BUBBLE_RADIUS, BUBBLE_RADIUS, Path.Direction.CW);
        }
        path.close();
        whitePaint.setColor(color);
        whitePaint.setStyle(Paint.Style.FILL);
        canvas.drawPath(path, whitePaint);
    }

    private int dp(float v, Context context) {
        return (int) (v * context.getResources().getDisplayMetrics().density + 0.5);
    }
}

File: app/src/main/java/com/tomclaw/appsend/util/BubbleColorDrawable.kt
package com.tomclaw.appsend.util

import android.content.Context
import android.graphics.Canvas
import android.graphics.ColorFilter
import android.graphics.Paint
import android.graphics.Path
import android.graphics.PixelFormat
import android.graphics.RectF
import android.graphics.drawable.Drawable

enum class Corner {
    LEFT, RIGHT, NONE
}

class BubbleColorDrawable(
    context: Context,
    private val color: Int,
    private val corner: Corner
) : Drawable() {

    private val whitePaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var offset = dp(8f, context).toFloat()
    private var bubbleRadius = dp(6f, context).toFloat()

    override fun setColorFilter(cf: ColorFilter?) {}

    override fun setAlpha(alpha: Int) {}

    @Deprecated(
        "Deprecated in Java",
        ReplaceWith("PixelFormat.OPAQUE", "android.graphics.PixelFormat")
    )
    override fun getOpacity(): Int {
        return PixelFormat.OPAQUE
    }

    override fun draw(canvas: Canvas) {
        val r = RectF(bounds)
        val rect = RectF(r)
        rect.inset(offset, offset)
        val path = Path()
        when (corner) {
            Corner.LEFT -> {
                path.moveTo(0f, 0f)
                path.lineTo(r.width() - offset, 0f)
                path.arcTo(
                    RectF(
                        r.right - bubbleRadius * 2f,
                        0f,
                        r.right,
                        bubbleRadius * 2f
                    ), 270f, 90f
                )
                path.lineTo(r.width(), r.height() - offset)
                path.arcTo(
                    RectF(
                        r.right - bubbleRadius * 2f,
                        r.bottom - bubbleRadius * 2f,
                        r.right,
                        r.bottom
                    ),
                    0f,
                    90f
                )
                path.lineTo(bubbleRadius, r.height())
                path.arcTo(
                    RectF(
                        offset,
                        r.bottom - bubbleRadius * 2f,
                        bubbleRadius * 2f + offset,
                        r.bottom
                    ),
                    90f,
                    90f
                )
                path.lineTo(offset, offset)
            }

            Corner.RIGHT -> {
                path.moveTo(bubbleRadius, 0f)
                path.lineTo(r.width(), 0f)
                path.lineTo(r.width() - offset, offset)
                path.lineTo(r.width() - offset, (r.height() - bubbleRadius))
                path.arcTo(
                    RectF(
                        r.right - bubbleRadius * 2f - offset,
                        r.bottom - bubbleRadius * 2f,
                        r.right - offset,
                        r.bottom
                    ),
                    0f,
                    90f
                )
                path.lineTo(bubbleRadius, r.height())
                path.arcTo(
                    RectF(
                        0f,
                        r.bottom - bubbleRadius * 2f,
                        bubbleRadius * 2f,
                        r.bottom
                    ), 90f, 90f
                )
                path.lineTo(0f, bubbleRadius)
                path.arcTo(
                    RectF(
                        0f,
                        0f,
                        bubbleRadius * 2f,
                        bubbleRadius * 2f
                    ),
                    180f,
                    90f
                )
            }

            Corner.NONE -> {
                rect.inset(-offset, -offset)
                path.addRoundRect(
                    rect,
                    bubbleRadius,
                    bubbleRadius,
                    Path.Direction.CW
                )
            }
        }
        path.close()
        whitePaint.color = color
        whitePaint.style = Paint.Style.FILL
        canvas.drawPath(path, whitePaint)
    }

    private fun dp(v: Float, context: Context): Int {
        return (v * context.resources.displayMetrics.density + 0.5).toInt()
    }
}
